lib/activerecord_autoreplica.rb

```diff
@@ -69,32 +88,73 @@ def self.using_read_replica_pool(replica_connection_pool)
     in_replica_context(replica_connection_pool){ yield }
   end
 
-  def self.in_replica_context(handler_params, handler_class=ConnectionHandler)
-    return yield if Thread.current[:autoreplica] # This method should not be reentrant
+  def self.in_replica_context(read_pool)
+    return yield if current_read_pool # This method should not be reentrant
+
+    # There is a pontential race condition here in a threaded environment, but
+    # in the worst case the handler will be set up twice. This shouldn't affect
+    # operations.
```

Comment:
Well, I never considered a mutex to be cheap; although I agree it might not make a big difference here. But in general, I'd always prefer to not have a mutex if it's avoidable... No big deal though, happy to add one if upstream requests it.